0000-interledger-protocol-4/0000-interledger-protocol-4.md

```diff
@@ -0,0 +1,173 @@
+
+## Flow
+
+Before the protocol starts, the sender and receiver use a higher-level protocol to agree on the destination account and packet condition.
+
+1. Sender constructs an ILP Prepare packet with the receiver's destination account, the agreed-upon condition, and an amount and expiry of their choice. The sender may include additional data, the formatting of which is determined by the higher-level protocol.
+2. Sender sends the ILP Prepare packet to a connector over some communication channel, such as a Websocket. 
+3. Connector gets the packet and determines who the sender was based on the communication channel it was received through. The connector checks whether the sender has sufficient balance or credit to send the amount specified in the packet. If so, the connector reduces the sender's available balance by the value of the packet. If not, the connector returns an ILP Reject packet.
+4. Connector uses its local routing tables and the destination ILP address from the packet to determine which next hop to forward the packet to.
+5. Connector modifies the packet amount and expiry to apply their exchange rate and move the expiry timestamp earlier (for example, each connector may subtract one second from the expiry to give themselves a minimum of one second to pass the Fulfill on in step 9). 
+6. Connector forwards the packet to the next hop, which may be another connector. All subsequent connectors go through steps 3-6 (treating the previous connector as the sender) until the packet reaches the receiver.
+7. Receiver checks the ILP Prepare packet, according to whatever is stipulated by the higher-level protocol (such as checking whether the amount received is above some minimum specified by the sender). Receiver also checks that the Prepare would not put the connector's unsettled balance above the maximum the receiver tolerates (this is the same as the minimum balance limit of the connector _with the receiver_).
+8. To accept the packet, the receiver returns an ILP Fulfill packet, containing the preimage of the condition in the ILP Prepare packet. The receiver may include additional data, the formatting of which is determined by the higher-level protocol. If the receiver does not want the ILP Prepare packet or it does not pass one of the checks, the receiver returns an ILP Reject packet.
```

Comment:
On the same connection that the connector received the ILP Prepare packet on?
If the connection is lost, the connector may not be able to  return the fulfillment packet because they can't identify who the sender was (ILP packet doesn't contain sender address).
Am I missing something?

Reply:
The connection has to be authenticated in such a way that the connector can identify the sender and recognize which pre-established account¹ is theirs.

¹ In this case, the "account" is probably a payment channel but it could also be some other means of tracking a balance between the connector and the sender.

Reply:
@mDuo13 is there protocol for it?